File: golang/Problem276.go
package main

import (
	"euler"
	"fmt"
	"time"
)

// Number of triangles
func Triangle(n int64) int64 {
	return ((n * n / 6) + 1) / 2
}

// Number of primitive triangles
func Primitive(n int64) int64 {
	if euler.IsPrime(n) {
		return Triangle(n)
	}

	ans := int64(0)
	for i := int64(2); i < n; i++ {
		if n%i == 0 {
			ans += Primitive(i)
		}
	}
	return Triangle(n) - ans

}

func main() {
	starttime := time.Now()

	for i := int64(3); i < 10; i++ {
		fmt.Println(i, Triangle(i), Primitive(i))
	}

	fmt.Println("Elapsed time:", time.Since(starttime))
}

File: python/Problem276.py
# Number of triangles
def Triangle(n):
    return ((n * n / 6) + 1) / 2 - (n / 4) * ((n + 2) / 4)


def seive(top):
    seive_data = [0 for i in xrange(top)]

    accumulate = 0

    for n in xrange(2, top):
        # if prime
        if seive_data[n] == 0:

            seive_data[n] = Triangle(n)
        else:
            seive_data[n] = Triangle(n) - seive_data[n]

        delta = seive_data[n]
        accumulate += delta

        k = n + n
        while k < top:
            seive_data[k] += delta
            k += n

    return accumulate


print seive(1 + 10 ** 7)
